Simplify private npm setup (#1858)
seek-oss/skuba: docs/deep-dives/npm.md

**skuba** advises migrating to `pnpm` if possible.

`docker-ecr-cache` is very commonly used in SEEK **skuba** projects.
Some **skuba** templates, like `express-koa-api`, come with a `docker-ecr-cache` setup that has this included.
This will look like:

```yaml
# .buildkite/pipeline.yml
configs:
  plugins:
    - &docker-ecr-cache
      seek-oss/docker-ecr-cache#v2.2.1:
        cache-on:
          - .npmrc
          - package.json#.packageManager
          - pnpm-lock.yaml
        dockerfile: Dockerfile.dev-deps
        secrets:
          # This mounts the provisioned .npmrc file and the NPM_TOKEN environment variable for use in `pnpm fetch`
          - id=npm,src=/var/lib/buildkite-agent/.npmrc
          - NPM_TOKEN

steps:
  - label: Some step
    commands:
      # Private npm access is not needed here; it is only used by `pnpm fetch` in the Dockerfile
      - pnpm install --offline
      - echo "Hello world"
    env:
      # This tells our Buildkite setup to create the .npmrc file and provision the NPM_TOKEN environment variable
      GET_NPM_TOKEN: please
    plugins:
      - *docker-ecr-cache
      - docker-compose#v5.6.0:
          run: app
          environment:

>             - GITHUB_API_TOKEN

Reviewer: Is it worth spelling out that you _shouldn't_ pass through `NPM_TOKEN` here?

Author: ```suggestion
    env:
      # This tells our Buildkite setup to create the .npmrc file and provision the NPM_TOKEN environment variable
      GET_NPM_TOKEN: please
    plugins:
      - *docker-ecr-cache
      - docker-compose#v5.6.0:
          run: app
          environment:
            # Don't pass NPM_TOKEN to the container. It's only needed in docker-ecr-cache.
            - GITHUB_API_TOKEN
```

   ```diff
     seek-oss/docker-ecr-cache#v2.2.1:
       cache-on:
         - .npmrc
         - package.json#.packageManager
         - pnpm-lock.yaml
       dockerfile: Dockerfile.dev-deps
   -   secrets: id=npm,src=/tmp/.npmrc
   +   secrets:
   +     - id=npm,src=/var/lib/buildkite-agent/.npmrc
   +     - NPM_TOKEN
   ```

3. Add the `GET_NPM_TOKEN` environment variable to any step which uses `docker-ecr-cache`.

   ```diff
     steps:
       - label: Some step
         plugins:
          - *docker-ecr-cache
   +     env:
   +       GET_NPM_TOKEN: please
   ```

   🚨 Be wary of any steps which retrieve `env` variables via yaml anchors.

   An example of a mistake:

   ```yaml
   configs:
     prod: &prod
       plugins: [...]
       env:
         ENVIRONMENT: production

   steps:
     - label: Some step
       <<: *prod
       env:
         # 🚨🚨🚨 This will override the ENVIRONMENT variable from the prod anchor
         GET_NPM_TOKEN: please
   ```

   In order to avoid this, you could put `GET_NPM_TOKEN` in the `configs` section too, or set `GET_NPM_TOKEN`
   as a pipeline-wide environment variable.

   ```yaml
   env:
     GET_NPM_TOKEN: please

   steps: ...

>    ```

Reviewer: Is this an antipattern? I wonder if the advice should be to reduce the anchor indirection (i.e. just move the environment variables inline).

Author: I think it is, because it incurs cost to load

Author: How about:

```suggestion
   In order to avoid this, you could: 
     - Reduce indirection with the YAML anchors, and put all environment variables inline in the step
     - Put `GET_NPM_TOKEN` in the `configs` section too
     - Set `GET_NPM_TOKEN` as a pipeline-wide environment variable, incurring additional startup cost on any steps that don't need it
```

Reviewer: I'd personally prefer to remove the last point but otherwise lgtm 🙂

>    4. Update your Dockerfile to use the new secrets.
> 
>    ```diff
>       RUN --mount=type=bind,source=.npmrc,target=.npmrc \
>           --mount=type=bind,source=package.json,target=package.json \
>           --mount=type=bind,source=pnpm-lock.yaml,target=pnpm-lock.yaml \
>           --mount=type=secret,id=npm,dst=/root/.npmrc,required=true \
>     +     --mount=type=secret,id=NPM_TOKEN,env=NPM_TOKEN,required=true \
>     +     pnpm fetch
>    ```

Reviewer: ````suggestion
4. Update your Dockerfile to use the new secrets.

   ```diff
     RUN --mount=type=bind,source=.npmrc,target=.npmrc \
         --mount=type=bind,source=package.json,target=package.json \
         --mount=type=bind,source=pnpm-lock.yaml,target=pnpm-lock.yaml \
         --mount=type=secret,id=npm,dst=/root/.npmrc,required=true \
   +     --mount=type=secret,id=NPM_TOKEN,env=NPM_TOKEN,required=true \
   +     pnpm fetch
   ```
````

---

## Other setups

Other setups are likely to look similar, but are not covered here.

> Review the [internal npm documentation](https://backstage.myseek.xyz/docs/default/component/artifact-management-docs/npm-access/).

Author: ```suggestion
Review the [internal npm documentation](https://backstage.myseek.xyz/docs/default/component/artifact-management-docs/npm/access/).
```
